lib/SIL/TypeLowering.cpp

```diff
@@ -2520,3 +2520,69 @@ TypeConverter::getContextBoxTypeForCapture(ValueDecl *captured,
   
   return boxType;
 }
+
+size_t Lowering::countNumberOfFields(SILModule &Module, SILType Ty) {
+  size_t fieldsCount = 0;
+  CanType canType = Ty.getSwiftRValueType();
+
+  std::function<bool(Type)> traverseAndCount = [&](Type type) -> bool {
+    if (auto *structDecl = type->getStructOrBoundGenericStruct()) {
+      for (auto prop : structDecl->getStoredProperties()) {
+        if (!prop->hasType()) {
+          continue;
+        }
+        auto propTy = prop->getType();
+        size_t fieldsCountBefore = fieldsCount;
+        propTy.findIf(traverseAndCount);
+        if (fieldsCount == fieldsCountBefore) {
+          // size of Struct(BigStructType) == sizeo of BigStructType()
+          // prevent counting its size as BigStructType()+1
+          ++fieldsCount;
+        }
+      }
+    } else if (type->is<TupleType>()) {
+      auto tupleTy = type->castTo<TupleType>();
+      for (auto elt : tupleTy->getElementTypes()) {
+        auto canElt = elt->getCanonicalType();
+        canElt.findIf(traverseAndCount);
+      }
+    } else if (auto *enumDecl = type->getEnumOrBoundGenericEnum()) {
+      if (enumDecl->classifyAsOptionalType()) {
+        // Special case:
+        // The traverser will handle the optional payload on its own
+        return false;
+      }
+      auto currSILType =
+          SILType::getPrimitiveObjectType(type->getCanonicalType());
+      size_t fieldsCountBefore = fieldsCount;
+      size_t maxEnumCount = 0;
+      for (auto elt : enumDecl->getAllElements()) {
+        if (!elt->hasAssociatedValues()) {
+          continue;
+        }
+        auto eltTy = currSILType.getEnumElementType(elt, Module);
+        auto eltCanTy = eltTy.getSwiftRValueType();
+        fieldsCount = 0;
+        eltCanTy.findIf(traverseAndCount);
+        if (fieldsCount > maxEnumCount) {
+          maxEnumCount = fieldsCount;
+        }
+      }
+      fieldsCount = fieldsCountBefore + maxEnumCount;
+    }
+    return true;
+  };
+
+  canType.findIf(traverseAndCount);
+  return std::max(fieldsCount, 1UL);
```

Comment:
You might want to cache the result of this computation, since it's performed in a lot of places